### PR TITLE
feat(vmcpu): added vmcpu resource and controller

### PR DESCRIPTION
**Pull request by Isteb4k**
## Description
Added VirtualMachineCPUModel (vmcpu) resource and controller
Refactored vmattachee - moved the disk/image/cpu-specific code from the vmattachee package (open for extensions, closed for modifications principle)

## Why do we need it, and what problem does it solve?
PR provides ability to create virtual machines with the required virtual CPU model

## What is the expected result?
Resources in cluster: virtualmachinecpumodel.virtualization.deckhouse.io
Controller works with this resource
Ability to bound vmcpu to virtual machine throw vm.spec.cpu.model

## Checklist
- [ ] The code is covered by unit tests.
- [X] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	var features virtv2.VirtualMachineCPUModelStatusFeatures
+
+	for feature, nodeCount := range enabledFeatures {
+		if nodeCount != len(nodes) {
+			continue
+		}
+
+		features.Enabled = append(
+			features.Enabled,
+			strings.TrimPrefix(feature, v1.CPUFeatureLabel),
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go:130]: kubevirt label

File: templates/virtualization-controller/vmcpu-legacy.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: virtualization.deckhouse.io/v1alpha2
+kind: VirtualMachineCPUModel
+metadata:
+  name: legacy
+spec:
+  type: Model
+  model: "Nehalem"
```
Comment (yaroslavborbat) [inline templates/virtualization-controller/vmcpu-legacy.yaml:7]: perhaps we want to create all available cpu models at once.  https://github.com/libvirt/libvirt/blob/master/src/cpu_map/index.xml

File: images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		modelNodes := r.getModelNodes(state.VMCPU.Current().Spec.Model, state.Nodes)
+		state.VMCPU.Changed().Status.Nodes = &modelNodes
+	}
+
+	if state.VMCPU.Current().Spec.Type == virtv2.Features {
+		features := r.getCommonNodeFeatures(state.Nodes)
+		state.VMCPU.Changed().Status.Features = &features
+	}
+
+	state.VMCPU.Changed().Status.Phase = virtv2.VMCPUPhaseReady
```
[the remark this answers is not in the document]
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go:70]: This is what we expect

File: templates/virtualization-controller/vmcpu-legacy.yaml (hunk 2 above)
Comment (Isteb4k) [reply templates/virtualization-controller/vmcpu-legacy.yaml:7]: It's necessary to create only Nehalem model as preset (in accordance with rnd/api/cpu-model requirements)

File: crds/virtualmachine.yaml [hunk 4]
```diff
@@ -811,6 +811,11 @@ spec:
                   required:
                     - cores
                   properties:
+                    model:
+                      type: string
+                      default: "legacy"
```
Comment (diafour) [inline crds/virtualmachine.yaml:816]: I feel that `legacy` is not a correct default here. It tells user nothing about real default model, i.e. "Nehalem"

File: images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+func NewVMCPUValidator(log logr.Logger) *VMCPUValidator {
+	return &VMCPUValidator{log: log}
+}
+
+type VMCPUValidator struct {
+	log logr.Logger
+}
+
+func (v *VMCPUValidator) ValidateCreate(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	err := fmt.Errorf("misconfigured webhook rules: delete operation not implemented")
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go:25]:
```suggestion
	err := fmt.Errorf("misconfigured webhook rules: create operation not implemented")
```

File: images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+	if !slices.Equal(newVMCPU.Spec.Features, oldVMCPU.Spec.Features) {
+		return nil, errors.New("virtual machine CPU features cannot be changed once set")
+	}
+
+	return nil, nil
+}
+
+func (v *VMCPUValidator) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	err := fmt.Errorf("misconfigured webhook rules: delete operation not implemented")
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go:57]: Shouldn't we forbid the deletion of "default" resource? (Despite its name)

File: images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+		return nil
+	}
+
+	state.VMCPU.Changed().Status.Phase = virtv2.VMCPUPhasePending
+
+	return nil
+}
+
+func contains(src []string, items ...string) bool {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/cpu/cpu_reconciler.go:118]: It should have a comment or rename it to the "containAll".

File: images/virtualization-artifact/pkg/controller/vm_reconciler_test.go [hunk 8]
```diff
@@ -374,6 +392,7 @@ var _ = Describe("Apply VM changes", func() {
 		nsName := "test-ns-2"
 		vmName := "test-vm-2"
 		vmipName := "test-vmip"
+		vmcpuName := "test-vmip"
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vm_reconciler_test.go:395]:
```suggestion
		vmcpuName := "test-vmcpu"
```
Just for consistency =)

**Review by diafour — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go (hunk 6 above)
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/cpu/cpu_webhook.go:57]: No, the user should be able to delete the default resource.

**Review by diafour — APPROVED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**